File: libs/power-log-parser/src/lib/handlers/data-handler.ts
import type { Helper } from '../helper';
import type { GameMetaData } from '../state/game-meta-data';
import type { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

import { BlockEndHandler } from './block-end-handler';
import { BlockStartHandler } from './block-start-handler';
import { ChangeEntityHandler } from './change-entity-handler';
import { FullEntityHandler } from './full-entity-handler';
import { HideEntityHandler } from './hide-entity-handler';
import { MetaDataHandler } from './meta-data-handler';
import { NewGameHandler } from './new-game-handler';
import { ShowEntityHandler } from './show-entity-handler';
import { SpectatorHandler } from './spectator-handler';
import { SubSpellHandler } from './sub-spell-handler';
import { TagChangeHandler } from './tag-change-handler';
import { TagHandler } from './tag-handler';
import { ShuffleDeckHandler } from './shuffle-deck-handler';
import { CreateGameHandler } from './create-game-handler';
import { CreatePlayerHandler } from './create-player-handler';
import { PlayerNameHandler } from './player-name-handler';
import { ActionMetadataHandler } from './action-metadata-handler';
import { ActionMetadataInfoHandler } from './action-metadata-info-handler';

type HandlerFn = (
	timestamp: string,
	data: string,
	state: ParserState,
	stateType: StateType,
	facade: StateFacade,
	indent: number,
) => boolean;

export class DataHandler {
	private readonly helper: Helper;
	private readonly metadata: GameMetaData = { BuildNumber: 0, FormatType: 0, GameType: 0, ScenarioID: 0 };
	private readonly handlers: [string, HandlerFn][];

	constructor(helper: Helper) {
		this.helper = helper;

		this.handlers = [
			['GameEntity EntityID=', (ts, d, s, st, f, i) => CreateGameHandler.Handle(ts, d, s, i)],
			['PlayerID=', (ts, d, s, st, f, i) => PlayerNameHandler.Handle(ts, d, s, st)],
			['Player EntityID=', (ts, d, s, st, f, i) => CreatePlayerHandler.Handle(d, s, f, i)],
			['BLOCK_START', (ts, d, s, st, f, i) => BlockStartHandler.Handle(ts, d, s, i, this.helper)],
			['BLOCK_END', (ts, d, s, st, f, i) => BlockEndHandler.Handle(d, s)],
			['BuildNumber=', (ts, d, s, st, f, i) => MetaDataHandler.Handle(ts, d, s, st, this.metadata, this.helper)],
			['GameType=', (ts, d, s, st, f, i) => MetaDataHandler.Handle(ts, d, s, st, this.metadata, this.helper)],
			['FormatType=', (ts, d, s, st, f, i) => MetaDataHandler.Handle(ts, d, s, st, this.metadata, this.helper)],
			['ScenarioID=', (ts, d, s, st, f, i) => MetaDataHandler.Handle(ts, d, s, st, this.metadata, this.helper)],
			['TAG_CHANGE', (ts, d, s, st, f, i) => TagChangeHandler.Handle(ts, d, s, st, f, i, this.helper)],
			['tag=', (ts, d, s, st, f, i) => TagHandler.Handle(ts, d, s, this.helper)],
			['SHUFFLE_DECK', (ts, d, s, st, f, i) => ShuffleDeckHandler.Handle(ts, d, s, i)],
			['FULL_ENTITY', (ts, d, s, st, f, i) => FullEntityHandler.Handle(ts, d, s, i, this.helper)],
			['SHOW_ENTITY', (ts, d, s, st, f, i) => ShowEntityHandler.Handle(ts, d, s, i, this.helper)],
			['CHANGE_ENTITY', (ts, d, s, st, f, i) => ChangeEntityHandler.Handle(ts, d, s, i, this.helper)],
			['HIDE_ENTITY', (ts, d, s, st, f, i) => HideEntityHandler.Handle(ts, d, s, i, this.helper)],
			['SUB_SPELL_START', (ts, d, s, st, f, i) => SubSpellHandler.Handle(ts, d, s, st, f, this.helper)],
			['Source =', (ts, d, s, st, f, i) => SubSpellHandler.Handle(ts, d, s, st, f, this.helper)],
			['Targets', (ts, d, s, st, f, i) => SubSpellHandler.Handle(ts, d, s, st, f, this.helper)],
			['SUB_SPELL_END', (ts, d, s, st, f, i) => SubSpellHandler.Handle(ts, d, s, st, f, this.helper)],
			['META_DATA', (ts, d, s, st, f, i) => ActionMetadataHandler.Handle(ts, d, s, i, this.helper)],
			['Info', (ts, d, s, st, f, i) => ActionMetadataInfoHandler.Handle(ts, d, s, i, this.helper)],
		];
	}

	Handle(
		timestamp: string,
		data: string,
		state: ParserState,
		stateType: StateType,
		previousTimestamp: string,
		stateFacade: StateFacade,
		currentGameSeed: number,
		resettingGame: boolean,
	): void {
		const trimmed = data.trim();
		const indentLevel = data.length - trimmed.length;
		data = trimmed;

		if (NewGameHandler.Handle(timestamp, data, state, previousTimestamp, stateType, stateFacade, currentGameSeed, resettingGame, this.metadata, this.helper)) {
			return;
		} else if (data.includes('Begin Spectating') || data.includes('End Spectator Mode')) {
			SpectatorHandler.Handle(timestamp, data, state, stateFacade);
			return;
		}

		if (state.Node != null) {
			for (const [key, handler] of this.handlers) {
				if (data.startsWith(key)) {
					if (handler(timestamp, data, state, stateType, stateFacade, indentLevel)) {
						return;
					}
				}
			}
		}

		if (state.Ended || state.CurrentGame == null) {
			return;
		}
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Parser/Handlers/DataHandler.cs
#region

using System;
using System.Collections.Generic;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.Entities;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using HearthstoneReplays.Parser.ReplayData.Meta;
using HearthstoneReplays.Events;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;
using Newtonsoft.Json;
using HearthstoneReplays.Parser.Handlers;

#endregion

namespace HearthstoneReplays.Parser.Handlers
{
    public class DataHandler
    {
        private readonly Helper helper;
        private readonly GameMetaData metadata = new GameMetaData();

        // Dictionary to map keywords to their respective handlers
        private readonly Dictionary<string, Func<DateTime, string, ParserState, StateType, StateFacade, int, bool>> handlers;

        public DataHandler(Helper helper)
        {
            this.helper = helper;

            // Initialize the handler dictionary
            handlers = new Dictionary<string, Func<DateTime, string, ParserState, StateType, StateFacade, int, bool>>
            {
                { "GameEntity EntityID=", (timestamp, data, state, stateType, facade, indent) => CreateGameHandler.HandleCreateGame(timestamp, data, state, indent) },
                { "PlayerID=", (timestamp, data, state, stateType, facade, indent) => PlayerNameHandler.HandlePlayerName(timestamp, data, state, stateType) },
                { "Player EntityID=", (timestamp, data, state, stateType, facade, indent) => CreatePlayerHandler.HandleCreatePlayer(data, state, facade, indent) },
                { "BLOCK_START", (timestamp, data, state, stateType, facade, indent) => BlockStartHandler.HandleBlockStart(timestamp, data, state, indent, helper) },
                { "BLOCK_END", (timestamp, data, state, stateType, facade, indent) => BlockEndHandler.HandleBlockEnd(data, state) },
                { "BuildNumber=", (timestamp, data, state, stateType, facade, indent) => MetaDataHandler.HandleMetaData(timestamp, data, state, stateType, metadata, helper) },
                { "GameType=", (timestamp, data, state, stateType, facade, indent) => MetaDataHandler.HandleMetaData(timestamp, data, state, stateType, metadata, helper) },
                { "FormatType=", (timestamp, data, state, stateType, facade, indent) => MetaDataHandler.HandleMetaData(timestamp, data, state, stateType, metadata, helper) },
                { "ScenarioID=", (timestamp, data, state, stateType, facade, indent) => MetaDataHandler.HandleMetaData(timestamp, data, state, stateType, metadata, helper) },
                { "TAG_CHANGE", (timestamp, data, state, stateType, facade, indent) => TagChangeHandler.HandleTagChange(timestamp, data, state, stateType, facade, indent, helper) },
                { "tag=", (timestamp, data, state, stateType, facade, indent) => TagHandler.HandleTag(timestamp, data, state, helper) },
                { "SHUFFLE_DECK", (timestamp, data, state, stateType, facade, indent) => ShuffleDeckHandler.HandleShuffleDeck(timestamp, data, state, indent) },
                { "FULL_ENTITY", (timestamp, data, state, stateType, facade, indent) => FullEntityHandler.HandleFullEntity(timestamp, data, state, indent, helper) },
                { "SHOW_ENTITY", (timestamp, data, state, stateType, facade, indent) => ShowEntityHandler.HandleShowEntity(timestamp, data, state, indent, helper) },
                { "CHANGE_ENTITY", (timestamp, data, state, stateType, facade, indent) => ChangeEntityHandler.HandleChangeEntity(timestamp, data, state, indent, helper) },
                { "HIDE_ENTITY", (timestamp, data, state, stateType, facade, indent) => HideEntityHandler.HandleHideEntity(timestamp, data, state, indent, helper) },
                { "SUB_SPELL_START", (timestamp, data, state, stateType, facade, indent) => SubSpellHandler.HandleSubSpell(timestamp, data, state, stateType, facade, helper) },
                { "Source =", (timestamp, data, state, stateType, facade, indent) => SubSpellHandler.HandleSubSpell(timestamp, data, state, stateType, facade, helper) },
                { "Targets", (timestamp, data, state, stateType, facade, indent) => SubSpellHandler.HandleSubSpell(timestamp, data, state, stateType, facade, helper) },
                { "SUB_SPELL_END", (timestamp, data, state, stateType, facade, indent) => SubSpellHandler.HandleSubSpell(timestamp, data, state, stateType, facade, helper) },
                { "META_DATA", (timestamp, data, state, stateType, facade, indent) => ActionMetadataHandler.HandleActionMetaData(timestamp, data, state, indent, helper) },
                { "Info", (timestamp, data, state, stateType, facade, indent) => ActionMetadataInfoHandler.HandleActionMetaDataInfo(timestamp, data, state, indent, helper) },
            };
        }

        public void Handle(DateTime timestamp, string data, ParserState state, StateType stateType, DateTime previousTimestamp, StateFacade stateFacade, 
            long currentGameSeed, bool resettingGame)
        {
            var trimmed = data.Trim();
            //Logger.Log("trimmed", trimmed);
            var indentLevel = data.Length - trimmed.Length;
            data = trimmed;

            // Additional handlers for specific cases
            // Take care of leftover log lines from a possible previous game
            if (NewGameHandler.HandleNewGame(timestamp, data, state, previousTimestamp, stateType, stateFacade, currentGameSeed, resettingGame, metadata, helper))
            {
                return;
            }
            else if (data.Contains("Begin Spectating") || data.Contains("End Spectator Mode"))
            {
                SpectatorHandler.HandleSpectator(timestamp, data, state, stateFacade);
                return;
            }

            if (state.Node != null)
            {
                // Pre-filter lines using string operations
                foreach (var handler in handlers)
                {
                    if (data.StartsWith(handler.Key))
                    {
                        if (handler.Value(timestamp, data, state, stateType, stateFacade, indentLevel))
                        {
                            return;
                        }
                    }
                }
            }

            // Handle fallback cases
            if (state.Ended || state.CurrentGame == null)
            {
                return;
            }
        }
    }
}